Improve and update ResourceQuota for 1.3

derekwaynecarr opened the pull request:
This is the updates for quota for 1.3

@pmorie @ncdc - PTAL

derekwaynecarr commented:
@ncdc - poke for a quick review pass.

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -4,29 +4,26 @@
+  tracks usage to ensure it does not exceed hard resource limits defined in a Resource Quota.
+- If creating or updating a resource violates a quota constraint, the request will fail with HTTP
+  status code `403 FORBIDDEN` with a message explaining the constraint that would have been violated.
+- If quota is enabled in a namespace for compute resources like `cpu` and `memory`, user's must specify
```
users

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -4,29 +4,26 @@
+- If creating or updating a resource violates a quota constraint, the request will fail with HTTP
+  status code `403 FORBIDDEN` with a message explaining the constraint that would have been violated.
+- If quota is enabled in a namespace for compute resources like `cpu` and `memory`, user's must specify
+  requests or limits for those values otherwise the quota system may reject pod creation.  Hint: Use
```
values, otherwise

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -70,12 +68,15 @@ are supported:
+| pods | The total number of pods in a non-terminal state that can exist in the namespace.  A pod is in a terminal state if `status.phase in (Failed, Succeeded)` is true.  |
+| replicationcontrollers | The total number of replication controllers that can exist in the namespace. |
+| resourcequotas | The total number of [resource quotas](/docs/admin/admission-controllers/#resourcequota) that can exist in the namespace. |
+| services | The total number of serivces that can exist in the namespace. |
```
typo: serivces

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -70,12 +68,15 @@ are supported:
+| replicationcontrollers | The total number of replication controllers that can exist in the namespace. |
+| resourcequotas | The total number of [resource quotas](/docs/admin/admission-controllers/#resourcequota) that can exist in the namespace. |
+| services | The total number of serivces that can exist in the namespace. |
+| services.loadbalancers | The total number of serivces of type load balancer that can exist in the namespace. |
```
typo: serivces

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -70,12 +68,15 @@ are supported:
+| resourcequotas | The total number of [resource quotas](/docs/admin/admission-controllers/#resourcequota) that can exist in the namespace. |
+| services | The total number of serivces that can exist in the namespace. |
+| services.loadbalancers | The total number of serivces of type load balancer that can exist in the namespace. |
+| services.nodeports | The total number of serivces of type node port that can exist in the namespace. |
```
typo: serivces

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -84,45 +85,77 @@ You might want to set a pods quota on a namespace
+A `BestEffort` scope restricts a quota to limit the following resources:
+* pods
+
+A `Terminating`, `NotTerminating`, and `NotBestEffort` scope restricts a quota to tracking the following resources:
```
Do you think it'd sound better as "The `Terminating` ... scopes restrict a quota to tracking the following resources"? (same q for the previous para too.)

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -57,11 +54,12 @@ in a namespace can be limited.  The following compute resource types are support
-
-For example, `cpu` quota sums up the `resources.requests.cpu` fields of every
-container of every pod in the namespace, and enforces a maximum on that sum.
+| cpu | Across all pods in a non-terminal state, the sum of CPU requests cannot exceed this value. |
```
Is this `cpu` or `requests.cpu`?

derekwaynecarr replied:
`cpu` and `requests.cpu` are equivalent.  `cpu` was around before we had requests and limits.

johndmulhausen replied:
For each of the Resource Names, please wrap them in `backticks`

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -84,45 +85,77 @@ You might want to set a pods quota on a namespace
+    secrets: "10"
+    services: "10"
+    services.loadbalancers: "2"
 EOF
```
Add some newlines after here and some of the other lines below so it doesn't all run together?

derekwaynecarr replied:
done

ncdc commented on docs/admin/resourcequota/index.md:
```diff
@@ -84,45 +85,77 @@ You might want to set a pods quota on a namespace
-  },
-  "spec": {
-    "hard": {
-      "memory": "1Gi",
```
Do we not want to show memory and cpu any more?

derekwaynecarr replied:
I am trying to encourage using multiple quota documents to split compute-resource quotas from object count quotas.  I have a separate PR that adds an example of that, but I think for this scenario, I will add the compute resources section.

derekwaynecarr commented:
@ncdc - I think I addressed all your comments.  Thanks for the review, mind applying tech-lgtm so we can get this out the door before I am on vacation next week?

derekwaynecarr commented:
This PR now added the example (with updates to show new 1.3 features for services.*) quotas.

So this is the same content from https://github.com/kubernetes/kubernetes.github.io/pull/622 but with those tweaks added in 1.3.

johndmulhausen commented on docs/admin/resourcequota/index.md:
```diff
@@ -57,11 +54,12 @@ in a namespace can be limited.  The following compute resource types are support
 
 | ResourceName | Description |
```
Space between Resource and Name

johndmulhausen commented on docs/admin/resourcequota/index.md:
```diff
@@ -4,29 +4,26 @@
+- If quota is enabled in a namespace for compute resources like `cpu` and `memory`, users must specify
+  requests or limits for those values; otherwise, the quota system may reject pod creation.  Hint: Use
+  the LimitRange admission controller to force defaults for pods that make no compute resource requirements.
+  See the [walkthrough](/docs/admin/resourcequota/walkthrough.md) for an example to avoid this problem.
```
Take .md out of the URL and replace with /

johndmulhausen commented on docs/admin/resourcequota/index.md:
```diff
@@ -84,45 +85,107 @@ You might want to set a pods quota on a namespace
+* pods
+
+The `Terminating`, `NotTerminating`, and `NotBestEffort` scopes restrict a quota to tracking the following resources:
+* cpu
```
1 line of space btx text and list plz

ncdc commented on docs/admin/resourcequota/walkthrough.md:
```diff
@@ -1,164 +1,358 @@
+## Scenario
+
+The cluster-admin is operating a cluster on behalf of a user population and the cluster-admin
+wants to control the amount of resource that can be consumed in a particular namespace to promote
```
should this be "amount of resources"?

derekwaynecarr replied:
ack

ncdc commented on docs/admin/resourcequota/walkthrough.md:
```diff
@@ -1,164 +1,358 @@
+fair sharing of the cluster and control cost.
+
+The cluster-admin has the following goals:
+* Limit the amount of compute resource for running pods
```
resources?

ncdc commented on docs/admin/resourcequota/walkthrough.md:
```diff
@@ -1,164 +1,358 @@
 
-The Kubernetes API server is rejecting the ReplicaSet requests to create a pod because our pods
-do not specify any memory usage *request*.
+The Kubernetes API server is rejecting the replica set requests to create a pod because our pods
```
I think this might sound better as "The Kubernetes API server is rejecting the replica set's request to create a pod ..."

derekwaynecarr replied:
done.

ncdc commented:
A few questions/comments; otherwise, LGTM. I'll apply tech-lgtm once updates are pushed.

derekwaynecarr commented:
@ncdc @johndmulhausen - updates made

johndmulhausen commented on docs/admin/resourcequota/walkthrough.md:
```diff
@@ -1,164 +1,360 @@
-Note that `kubectl` commands will print the type and name of the resource created or mutated, which can then be used in subsequent commands: 
+## Step 2: Apply an object-count quota to the namespace
+
+The cluster-admin wants to control the following resources:
```
Blank space between text and list

johndmulhausen commented:
One last nit and you're good from my end :) Nice work, @derekwaynecarr !

ncdc commented:
A couple of small comments/questions

derekwaynecarr commented:
@johndmulhausen - last nit addressed.

johndmulhausen commented:
Thanks a ton, @derekwaynecarr !
